File: go_client/creds.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/google/uuid"
)

var vkSemaphore = make(chan struct{}, 2)

var (
	sharedTransportOnce sync.Once
	sharedTransport     *http.Transport
)

var noDnsFlag atomic.Bool

func getSharedTransport() *http.Transport {
	sharedTransportOnce.Do(func() {
		dialer := &net.Dialer{
			Timeout: 10 * time.Second,
		}
		sharedTransport = &http.Transport{
			Proxy:                 http.ProxyFromEnvironment,
			DialContext:           dialer.DialContext,
			ForceAttemptHTTP2:     true,
			MaxIdleConns:          100,
			MaxIdleConnsPerHost:   10,
			IdleConnTimeout:       90 * time.Second,
			TLSHandshakeTimeout:   10 * time.Second,
			ExpectContinueTimeout: 1 * time.Second,
		}
	})
	return sharedTransport
}

type Credentials struct {
	User     string
	Pass     string
	TurnURLs []string
}

func GetCredsWithFallback(ctx context.Context, tp *TurnParams, hash string, stats *Stats) (*Credentials, error) {
	creds, err := getUniqueVKCreds(ctx, hash, 5, stats)
	if err == nil {
		return creds, nil
	}
	if tp.SecondaryHash != "" && hash != tp.SecondaryHash {
		log.Println("Основной хеш не сработал, пробую запасной")
		return getUniqueVKCreds(ctx, tp.SecondaryHash, 3, stats)
	}
	return nil, err
}

func getUniqueVKCreds(ctx context.Context, hash string, maxRetries int, stats *Stats) (*Credentials, error) {
	var lastErr error

	for attempt := 0; attempt < maxRetries; attempt++ {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case vkSemaphore <- struct{}{}:
		}

		creds, err := getVKCredsOnce(ctx, hash)
		<-vkSemaphore

		if err == nil {
			return creds, nil
		}

		atomic.AddInt64(&stats.CredsErrors, 1)
		lastErr = err
		errStr := err.Error()

		if strings.Contains(errStr, "9000") || strings.Contains(errStr, "call not found") {
			return nil, fmt.Errorf("хеш мёртв: %w", err)
		}

		var backoff time.Duration
		if strings.Contains(errStr, "flood") || strings.Contains(errStr, "Flood") {
			secs := 5 * (attempt + 1)
			if secs > 60 {
				secs = 60
			}
			backoff = time.Duration(secs) * time.Second
		} else {
			base := 1 << uint(min(attempt, 5))
			if base > 30 {
				base = 30
			}
			backoff = time.Duration(base)*time.Second + time.Duration(rand.Intn(1000))*time.Millisecond
		}

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(backoff):
		}
	}

	return nil, fmt.Errorf("исчерпаны %d попыток: %w", maxRetries, lastErr)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func getVKCredsOnce(ctx context.Context, hash string) (*Credentials, error) {
	client := &http.Client{
		Timeout:   15 * time.Second,
		Transport: getSharedTransport(),
	}

	okAppKey := "CGMMEJLGDIHBABABA"
	appID := vkAppID.Load().(string)
	appSecret := vkAppSecret.Load().(string)

	doReq := func(data, url string) (map[string]interface{}, error) {
		req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBufferString(data))
		if err != nil {
			return nil, fmt.Errorf("создание запроса: %w", err)
		}
		req.Header.Set("User-Agent", "Mozilla/5.0")
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

		resp, err := client.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("чтение ответа: %w", err)
		}

		var m map[string]interface{}
		if err := json.Unmarshal(body, &m); err != nil {
			return nil, fmt.Errorf("парсинг JSON: %w | Body: %s", err, string(body))
		}
		if errObj, ok := m["error"]; ok {
			return nil, fmt.Errorf("API error: %v | Body: %s", errObj, string(body))
		}
		return m, nil
	}

	get := func(m map[string]interface{}, keys ...string) (string, error) {
		var cur interface{} = m
		for _, k := range keys {
			mm, ok := cur.(map[string]interface{})
			if !ok {
				return "", fmt.Errorf("path %q not found", k)
			}
			cur = mm[k]
		}
		s, ok := cur.(string)
		if !ok {
			return "", fmt.Errorf("value at path is not string")
		}
		return s, nil
	}

	// 1: get anonymous token (с profile scopes и appSecret)
	r, err := doReq(fmt.Sprintf(
		"client_secret=%s&client_id=%s&scopes=audio_anonymous%%2Cvideo_anonymous%%2Cphotos_anonymous%%2Cprofile_anonymous&isApiOauthAnonymEnabled=false&version=1&app_id=%s",
		appSecret, appID, appID,
	), "https://login.vk.ru/?act=get_anonym_token")
	if err != nil {
		return nil, fmt.Errorf("шаг 1: %w", err)
	}
	t1, err := get(r, "data", "access_token")
	if err != nil {
		return nil, fmt.Errorf("шаг 1 парсинг: %w", err)
	}

	// 2: get messages token используя payload (t1)!!! (Это было пропущено)
	r, err = doReq(fmt.Sprintf(
		"client_id=%s&token_type=messages&payload=%s&client_secret=%s&version=1&app_id=%s",
		appID, t1, appSecret, appID,
	), "https://login.vk.ru/?act=get_anonym_token")
	if err != nil {
		return nil, fmt.Errorf("шаг 2: %w", err)
	}
	t3, err := get(r, "data", "access_token")
	if err != nil {
		return nil, fmt.Errorf("шаг 2 парсинг: %w", err)
	}

	// 3: get call token используя t3
	r, err = doReq(fmt.Sprintf(
		"vk_join_link=https://vk.com/call/join/%s&name=123&access_token=%s",
		hash, t3,
	), "https://api.vk.ru/method/calls.getAnonymousToken?v=5.264")
	if err != nil {
		return nil, fmt.Errorf("шаг 3: %w", err)
	}
	t4, err := get(r, "response", "token")
	if err != nil {
		return nil, fmt.Errorf("шаг 3 парсинг: %w", err)
	}

	// 4: OK anonymous login
	r, err = doReq(fmt.Sprintf(
		"session_data=%%7B%%22version%%22%%3A2%%2C%%22device_id%%22%%3A%%22%s%%22%%2C%%22client_version%%22%%3A1.1%%2C%%22client_type%%22%%3A%%22SDK_JS%%22%%7D&method=auth.anonymLogin&format=JSON&application_key=%s",
		uuid.New().String(), okAppKey,
	), "https://calls.okcdn.ru/fb.do")
	if err != nil {
		return nil, fmt.Errorf("шаг 4: %w", err)
	}
	t5, err := get(r, "session_key")
	if err != nil {
		return nil, fmt.Errorf("шаг 4 парсинг: %w", err)
	}

	// 5: join conversation.
	r, err = doReq(fmt.Sprintf(
		"joinLink=%s&isVideo=false&protocolVersion=5&anonymToken=%s&method=vchat.joinConversationByLink&format=JSON&application_key=%s&session_key=%s",
		hash, t4, okAppKey, t5,
	), "https://calls.okcdn.ru/fb.do")
	if err != nil {
		return nil, fmt.Errorf("шаг 5: %w", err)
	}

	ts, ok := r["turn_server"].(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("turn_server не найден в ответе")
	}

	user, _ := ts["username"].(string)
	pass, _ := ts["credential"].(string)
	if user == "" || pass == "" {
		return nil, fmt.Errorf("пустые credentials в ответе")
	}

	lifetime, _ := ts["lifetime"].(float64)
	if lifetime > 0 {
		log.Printf("[ВК] Креды получены ✓ (LIFE: %.0f сек)", lifetime)
	} else {
		log.Printf("[ВК] Креды получены ✓")
	}

	urls, _ := ts["urls"].([]interface{})
	var turnAddrs []string
	for _, u := range urls {
		s, ok := u.(string)
		if !ok {
			continue
		}
		clean := strings.Split(s, "?")[0]
		addr := strings.TrimPrefix(strings.TrimPrefix(clean, "turn:"), "turns:")
		if addr != "" {
			turnAddrs = append(turnAddrs, addr)
		}
	}
	if len(turnAddrs) == 0 {
		return nil, fmt.Errorf("нет TURN urls в ответе")
	}

	return &Credentials{User: user, Pass: pass, TurnURLs: turnAddrs}, nil
}

File: src/creds.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use rand::Rng;
use tokio::sync::Semaphore;

use crate::config::TurnParams;
use crate::stats::Stats;

static VK_SEMAPHORE: tokio::sync::Semaphore = tokio::sync::Semaphore::const_new(2);

#[derive(Clone, Debug)]
pub struct Credentials {
    pub user: String,
    pub pass: String,
    pub turn_url: String,
}

pub async fn get_creds_with_fallback(
    tp: &TurnParams,
    hash: &str,
    shutdown: &Arc<AtomicBool>,
    stats: &Arc<Stats>,
    app_id: &str,
    app_secret: &str,
) -> Result<Credentials, String> {
    match get_unique_vk_creds(hash, 5, shutdown, stats, app_id, app_secret).await {
        Ok(c) => Ok(c),
        Err(e) => {
            if !tp.secondary_hash.is_empty() && hash != tp.secondary_hash {
                log::warn!("Основной хеш не сработал, пробую запасной");
                get_unique_vk_creds(&tp.secondary_hash, 3, shutdown, stats, app_id, app_secret)
                    .await
            } else {
                Err(e)
            }
        }
    }
}

async fn get_unique_vk_creds(
    hash: &str,
    max_retries: usize,
    shutdown: &Arc<AtomicBool>,
    stats: &Arc<Stats>,
    app_id: &str,
    app_secret: &str,
) -> Result<Credentials, String> {
    let mut last_err = String::new();

    for attempt in 0..max_retries {
        if shutdown.load(Ordering::Relaxed) {
            return Err("shutdown".into());
        }

        let _permit = VK_SEMAPHORE
            .acquire()
            .await
            .map_err(|e| format!("semaphore: {}", e))?;

        // Выполняем блокирующий HTTP в spawn_blocking
        let hash_owned = hash.to_string();
        let app_id_owned = app_id.to_string();
        let app_secret_owned = app_secret.to_string();

        let result = tokio::task::spawn_blocking(move || {
            get_vk_creds_once(&hash_owned, &app_id_owned, &app_secret_owned)
        })
        .await
        .map_err(|e| format!("spawn_blocking: {}", e))?;

        drop(_permit);

        match result {
            Ok(creds) => return Ok(creds),
            Err(e) => {
                stats.creds_errors.fetch_add(1, Ordering::Relaxed);
                last_err = e.clone();

                if e.contains("9000") || e.contains("call not found") {
                    return Err(format!("Хеш мёртв: {}", e));
                }

                let backoff = if e.contains("flood") || e.contains("Flood") {
                    let secs = std::cmp::min(5 * (attempt + 1), 60);
                    Duration::from_secs(secs as u64)
                } else {
                    let base = std::cmp::min(1u64 << std::cmp::min(attempt, 5), 30);
                    let jitter = rand::thread_rng().gen_range(0..1000);
                    Duration::from_secs(base) + Duration::from_millis(jitter)
                };

                if shutdown.load(Ordering::Relaxed) {
                    return Err("shutdown".into());
                }

                tokio::time::sleep(backoff).await;
            }
        }
    }

    Err(format!("Исчерпаны {} попыток: {}", max_retries, last_err))
}

fn get_vk_creds_once(hash: &str, app_id: &str, app_secret: &str) -> Result<Credentials, String> {
    let ok_app_key = "CGMMEJLGDIHBABABA";
    let timeout = Duration::from_secs(15);

    let agent = ureq::AgentBuilder::new()
        .timeout_connect(timeout)
        .timeout_read(timeout)
        .timeout_write(timeout)
        .user_agent("Mozilla/5.0")
        .build();

    let do_post =
        |url: &str, body: &str| -> Result<serde_json::Value, String> {
            let resp = agent
                .post(url)
                .set("Content-Type", "application/x-www-form-urlencoded")
                .send_string(body)
                .map_err(|e| format!("HTTP POST {}: {}", url, e))?;

            let text = resp
                .into_string()
                .map_err(|e| format!("Read response: {}", e))?;

            let val: serde_json::Value =
                serde_json::from_str(&text).map_err(|e| format!("JSON parse: {}", e))?;

            if val.get("error").is_some() {
                return Err(format!("API error: {}", val["error"]));
            }

            Ok(val)
        };

    let get_str = |val: &serde_json::Value, keys: &[&str]| -> Result<String, String> {
        let mut current = val;
        for &k in keys {
            current = current.get(k).ok_or_else(|| format!("Key '{}' not found", k))?;
        }
        current
            .as_str()
            .map(|s| s.to_string())
            .ok_or_else(|| "Value is not a string".into())
    };

    // Step 1: anonymous token
    let body1 = format!(
        "client_secret={}&client_id={}&scopes=audio_anonymous%2Cvideo_anonymous%2Cphotos_anonymous%2Cprofile_anonymous&isApiOauthAnonymEnabled=false&version=1&app_id={}",
        app_secret, app_id, app_id
    );
    let r1 = do_post("https://login.vk.ru/?act=get_anonym_token", &body1)?;
    let t1 = get_str(&r1, &["data", "access_token"])?;

    // Step 2: messages token
    let body2 = format!(
        "client_id={}&token_type=messages&payload={}&client_secret={}&version=1&app_id={}",
        app_id, t1, app_secret, app_id
    );
    let r2 = do_post("https://login.vk.ru/?act=get_anonym_token", &body2)?;
    let t2 = get_str(&r2, &["data", "access_token"])?;

    // Step 3: call token
    let body3 = format!(
        "vk_join_link=https://vk.com/call/join/{}&name=123&access_token={}",
        hash, t2
    );
    let r3 = do_post(
        "https://api.vk.ru/method/calls.getAnonymousToken?v=5.264",
        &body3,
    )?;
    let t3 = get_str(&r3, &["response", "token"])?;

    // Step 4: OK anonymous login
    let device_uuid = uuid::Uuid::new_v4();
    let body4 = format!(
        "session_data=%7B%22version%22%3A2%2C%22device_id%22%3A%22{}%22%2C%22client_version%22%3A1.1%2C%22client_type%22%3A%22SDK_JS%22%7D&method=auth.anonymLogin&format=JSON&application_key={}",
        device_uuid, ok_app_key
    );
    let r4 = do_post("https://calls.okcdn.ru/fb.do", &body4)?;
    let t4 = get_str(&r4, &["session_key"])?;

    // Step 5: join conversation
    let body5 = format!(
        "joinLink={}&isVideo=false&protocolVersion=5&anonymToken={}&method=vchat.joinConversationByLink&format=JSON&application_key={}&session_key={}",
        hash, t3, ok_app_key, t4
    );
    let r5 = do_post("https://calls.okcdn.ru/fb.do", &body5)?;

    let ts = r5
        .get("turn_server")
        .ok_or("turn_server не найден")?;

    let user = get_str(ts, &["username"])?;
    let pass = get_str(ts, &["credential"])?;

    if user.is_empty() || pass.is_empty() {
        return Err("Пустые credentials".into());
    }

    let urls = ts
        .get("urls")
        .and_then(|v| v.as_array())
        .ok_or("Нет TURN urls")?;

    if urls.is_empty() {
        return Err("Пустой массив TURN urls".into());
    }

    let turn_url_raw = urls[0].as_str().ok_or("TURN URL не строка")?;
    let clean = turn_url_raw.split('?').next().unwrap_or(turn_url_raw);
    let turn_addr = clean
        .trim_start_matches("turns:")
        .trim_start_matches("turn:")
        .to_string();

    Ok(Credentials {
        user,
        pass,
        turn_url: turn_addr,
    })
}